ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1295,4 +998,533 @@ private void endCompaction() {
+          }
+          if (currentFinalState.isDelete() && currentFinalState.getLifeVersion() == currentValue.getLifeVersion()) {
+            validEntries.add(entry);
+          }
+        } else if (currentValue.isTtlUpdate()) {
+          if (currentFinalState.isPut()) {
+            // If isPut returns true, when the final state doesn't carry ttl_update flag, this is wrong
+            throw new IllegalStateException(
+                "TtlUpdate's final state can't be put for key" + currentKey + " in store " + dataDir);
+          }
+          // This is a TTL_UPDATE record, then the blob can't be expired. Only check if it's deleted or not.
+          if (currentFinalState.isDelete()) {
+            if (currentFinalState.getOperationTimeInMs() >= compactionLog.getCompactionDetails().getReferenceTimeMs()) {
+              validEntries.add(entry);
+            } else if (isTtlUpdateEntryValidWhenFinalStateIsDeleteAndRetention(currentKey,
+                indexSegment.getStartOffset())) {
+              validEntries.add(entry);
+            }
+          } else {
+            validEntries.add(entry);
+          }
+        } else {
+          if (srcIndex.isExpired(currentFinalState)) {
+            logger.trace("{} in index segment with start offset {} in {} is not valid because it is an expired PUT",
+                entry, indexSegment.getStartOffset(), storeId);
+            continue;
+          }
+          if (currentFinalState.isPut()) {
+            if (currentFinalState.getLifeVersion() != currentValue.getLifeVersion()) {
+              throw new IllegalStateException(
+                  "Two different lifeVersions  for puts key" + currentKey + " in store " + dataDir);
+            }
+            validEntries.add(entry);
+          } else if (currentFinalState.isDelete()) {
+            if (currentFinalState.getOperationTimeInMs() >= compactionLog.getCompactionDetails().getReferenceTimeMs()) {
+              validEntries.add(entry);
+            } else {
+              logger.trace("{} in index segment with start offset {} in {} is not valid because it is a deleted PUT",
+                  entry, indexSegment.getStartOffset(), storeId);
+            }
+          } else {
+            validEntries.add(entry);
+          }
+        }
+      }
+      return validEntries;
+    }
+
+    /**
+     * Determines whether a TTL update entry is valid. A TTL update entry is valid as long as the associated PUT record
+     * is still present in the store (the validity of the PUT record does not matter - only its presence/absence does).
+     * @param key the {@link StoreKey} being examined
+     * @param indexSegmentStartOffset the start offset of the {@link IndexSegment} that the TTL update record is in
+     * @return {@code true} if the TTL update entry is valid
+     * @throws StoreException if there are problems reading the index
+     */
+    private boolean isTtlUpdateEntryValidWhenFinalStateIsDeleteAndRetention(StoreKey key,
```

Comment:
what does the retention mean, within retention time?

Reply:
Yes, current retention time is 7 days.

Reply:
final -> latest